dcos-log/api/v2/handlers.go

```diff
@@ -510,11 +545,65 @@ func journalHandler(w http.ResponseWriter, r *http.Request) {
 			{
 				_, err := io.Copy(w, j)
 				if err != nil {
-					logrus.Errorf("error while reading journald: %s. Request: %s", err, r.RequestURI)
+					logrus.Errorf("error while reading journald: %s. Request: %s", err, req.RequestURI)
 				}
 				f.Flush()
 			}
 		}
 	}
 
 }
+
+func browseFiles(w http.ResponseWriter, req *http.Request) {
+	r, err := setupFilesAPIReader(req, "/files/browse")
+	if err != nil {
+		e, ok := err.(errSetupFilesAPIReader)
+		if !ok {
+			logError(w, req, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		logError(w, req, e.msg, e.code)
+		return
+	}
+
+	files, err := r.BrowseSandbox()
+	if err != nil {
+		logError(w, req, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(files); err != nil {
+		logError(w, req, fmt.Sprintf("unable to encode sandbox files: %s. Items: %s", err, files), http.StatusInternalServerError)
+		return
+	}
+}
+
+func downloadFile(w http.ResponseWriter, req *http.Request) {
+	r, err := setupFilesAPIReader(req, "/files/download")
+	if err != nil {
+		e, ok := err.(errSetupFilesAPIReader)
+		if !ok {
+			logError(w, req, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		logError(w, req, e.msg, e.code)
+		return
+	}
+
+	downloadResp, err := r.Download()
+	if err != nil {
+		logError(w, req, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer downloadResp.Body.Close()
+
+	for k, vs := range downloadResp.Header {
+		for _, v := range vs {
+			w.Header().Add(k, v)
```

Comment:
Just curious... why are we adding all of the request headers into the response headers?  do we not want to just copy some of them?

Reply:
my idea was to give back the user the original response from mesos files API, which should contain the response headers. I did not find a good way to copy the entire header object, so i had to iterate over it